concept: generics

mpizenberg opened the pull request.

ceddlyburge commented on exercises/concept/story-name/.meta/design.md:
```diff
@@ -0,0 +1,28 @@
+## Out of scope
+
+- Covariance and contravariance.
+
```
Maybe add Never, Phantom Types and Opaque types as out of scope, although it would probably be useful to mention them in passing so that students know they exist

mpizenberg replied:
Well for Never, I'm not sure we even need to mention it since it's not really a generic, just a standard type that you have no way to construct. Phantom types, I'm not sure they are out of scope, at least introducing them. That will probably depend on if the concept grows a bit big already without them. As for opaque types, they are not linked to generics in any specific way, they are just normal types. The only thing special is that you don't export the types internal so it's more of public/private and module-related concept.

ceddlyburge replied:
Yep, that is true, although in my head they are all in the "advanced types" bucket :)

ceddlyburge commented on exercises/concept/story-name/.meta/design.md:
```diff
@@ -0,0 +1,28 @@
+
+## Learning objectives
+
+- Understand what a type variable is.
```
It might be worth calling out `Cmd Msg` and `Cmd msg` somewhere, which is something that confuses a lot of people new to elm I think.

mpizenberg replied:
Right! that's an important remark!

ceddlyburge commented on exercises/concept/story-name/.docs/introduction.md:
```diff
@@ -0,0 +1,27 @@
+## Higher order functions
+
+- basics of higher order functions: dependency injection.
+- common patterns accross container types (map, andThen, etc.)
```
Maybe add this to the learning objectives?

ceddlyburge commented on exercises/concept/story-name/.docs/introduction.md:
```diff
@@ -0,0 +1,27 @@
+# Introduction
+
+Generics are a way of writing code that avoid duplications of code.
```
```suggestion
Generics are a language feature that enable code reuse and advanced type patterns (which are explained in a later concept).
```

mpizenberg commented:
I've filled it and updated this with your remarks. I've also removed the section about higher order functions since it actually does not have anything to do with generics. This is a higher order function: `foo : (Int -> Int) -> String` for example.

Let me know what you think of the updated content!

ceddlyburge commented on exercises/concept/story-name/.docs/introduction.md:
```diff
@@ -1,27 +1,141 @@
+-- The module does not expose the internals of Distance,
+-- but instead two constructor functions: meter and foot.
+-- We do not want to expose 'Distance(..)' because we do not want
+-- people outside this module to access the internals of Distance.
```
I might word this bit slightly differently. The main thing with opaque types is that you can't create an instance of a type directly, you have to use one of the exported creator functions.

ceddlyburge reviewed (approved):
Looks good :)
